### ISSUE TITLE
Wrong error "TS2532: Object is possibly 'undefined'" after assigning to const with optional chaining

### ISSUE THREAD

**Issue by ssypi** [microsoft/TypeScript#40201]
**TypeScript Version:**  4.1.0-dev.20200823 also tested with 4.0.2

**Search Terms:** 
TS2532
Object is possibly 'undefined'
**Code**

```ts
interface Foo {
    foo: string;
    bar: string;
}

export default function getBar(obj: Foo | undefined): string {
    const foo = obj?.foo;
    if (!foo) {
        throw new Error("Invalid args");
    }

    return obj.bar; // <--- obj is possible undefined here according to tsc even though it can never really be
}

```

**Expected behavior:**

No typescript error should be shown here, because 'obj' can never be undefined since 'foo' would then also be undefined so the code would never reach the final return. 

**Actual behavior:**
with --strictNullChecks
`TS2532: Object is possibly 'undefined'` at the final return line for 'obj'

**Playground Link:**  
[Playground Link](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgGIHt3IN4ChkHIyYBcyAzmFKAOYDc+hARnFGZdSPbgL664QAHgAd0UMMgAmEeAFcANhJiyQCMMHQhkNCGABCrABTomAKzIYsAH2QrpMUBEkBKdlVo5GBBJspFMyAC8yCamAPwAdMToDITIwDDIhgCE0c6ecXFgABZQ6ADuyCAQhQCiUHlQhgBEAJIgAG5w8sCSyKw05NXOsYR8XshQurJQWqERLFB0yAD0M8gAPAC0KyFm8eTIouTkwEzyKHYyjm3Z0CiIPlCSHmBYYOQIyBANEFo56LI02fESCHBaYqvKCDCDNeQAT2QTAgvCAA)
**Related Issues:**

**Comment by ssypi** [comment]
There are ofcourse some workarounds, such as checking `if (!obj?.foo)` but if foo is used somewhere later and if there are more than one of these variables to check, it's not quite as good. And since the foo variable is even const here and the check is rather simple, I'd like to think the compiler should catch it correctly. Then again I know nothing about the internals so maybe it's not that simple in reality.
